msg.go

```diff
@@ -952,14 +967,29 @@ func (dns *Msg) String() string {
 			}
 		}
 	}
+	if len(dns.Stateful) > 0 {
+		s += "\n;; DSO SECTION:\n"
+		for _, v := range dns.Stateful {
+			s += fmt.Sprintf("%s: %s\n", dsoTypeToString(v.DSOType()), v)
+		}
+	}
+
 	return s
 }
 
 // isCompressible returns whether the msg may be compressible.
 func (dns *Msg) isCompressible() bool {
-	// If we only have one question, there is nothing we can ever compress.
-	return len(dns.Question) > 1 || len(dns.Answer) > 0 ||
-		len(dns.Ns) > 0 || len(dns.Extra) > 0
+	switch {
+	case dns.MsgHdr.Opcode == OpcodeStateful:
+		if isDSOCompressible(dns) {
+			return true
+		}
+		fallthrough
```

Comment:
This is done in spirit that if user wants to pack an invalid DSO message (with both DSO-Data and RR-data) we should not disallow compression.